### ISSUE TITLE
Integrating Homebrew with Homebrew Cask - GSOC

### ISSUE THREAD

**Issue by AnastasiaSulyagina** [Homebrew/Outreachy-and-Google-Summer-of-Code#6]
Hi @mikemcquaid !

I have a couple of new questions regarding GSOC application and proceeding with contributing to Homebrew.

- I was wondering if you could point me to some bug/feature/anything that could be fixed/done in ~week time? (Thought of doing something from [#15603](https://github.com/caskroom/homebrew-cask/issues/15603) as it has to be done for the project anyway, what do you think?) Guess it would be useful for me to become more familiar with language and Homebrew itself, moreover it could make my application more convincing.
- If I send my proposal draft for review, would you give me some feedback?

Thank you,
Anastasia

**Comment by miguelarauj1o** [comment]
If possible, the second question interesting me too.

**Comment by bfontaine** [comment]
Hello @AnastasiaSulyagina and @miguelarauj1o :smiley: 
The answer to the second question is yes: you’ll get feedback on your draft proposals.

**Comment by miguelarauj1o** [comment]
@bfontaine, thanks :)

**Comment by AnastasiaSulyagina** [comment]
Thank you, @bfontaine !

And one more silly question, what happens if smb applies for both GSOC and Outreachy? If one gets through only one program at time, what point in applying for both then? That does not increase chances anyway as you are either selected for a project or not, right? :D

**Comment by mikemcquaid** [comment]
@AnastasiaSulyagina Applying for both shouldn't make much of a difference to the students but it makes a big difference to the organisers as GSoC slots are paid by Google and Outreachy slots require the mentoring organisation to raise the funding somehow. Up to you though, of course, but as a mentor if you're eligible for both I'd probably try to get you into GSoC over Outreachy just to leave more Outreachy slots for people who aren't eligably for GSoC. Hope that makes sense!

**Comment by mikemcquaid** [comment]
(and just got off a :airplane: so will answer the other questions in more detail later)

**Comment by AnastasiaSulyagina** [comment]
@mikemcquaid Thank you, it makes sense now.

**Comment by mikemcquaid** [comment]
> I was wondering if you could point me to some bug/feature/anything that could be fixed/done in ~week time? (Thought of doing something from #15603 as it has to be done for the project anyway, what do you think?) Guess it would be useful for me to become more familiar with language and Homebrew itself, moreover it could make my application more convincing.

Yeh, that could be a good idea! Note that I'm not a maintainer on Homebrew Cask so we might want to talk to them about this stuff too. CC @jawshooah for thoughts here.

 Alternatively you could do something unrelated such as one of these https://github.com/homebrew/homebrew/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22

> If I send my proposal draft for review, would you give me some feedback?

@AnastasiaSulyagina @miguelarauj1o Yep, feel free to email me it at mike@mikemcquaid.com!

> That does not increase chances anyway as you are either selected for a project or not, right? :D

@AnastasiaSulyagina Also, I think it might increase your chances if you apply for both, actually!

**Comment by AnastasiaSulyagina** [comment]
Thank you again, @mikemcquaid ! 
I'll look at "help needed" things then and view some duplication-removing PRs to see how it is to be done properly for now.

**Comment by jawshooah** [comment]
@mikemcquaid I'm sure we'd be happy to have someone take on full-on integration as a GSOC project. Going to bring @vitorgalvao into this discussion as well.

**Comment by AnastasiaSulyagina** [comment]
Hi @jawshooah and @vitorgalvao !

As a part of my GSOC application I prepared work timeline for the project. Please view [the proposal](https://docs.google.com/document/d/1u6OZyfQMQz2JhvcV_Q4tAOHGgKXmpQqgQXn3RppwVMQ/edit?usp=sharing) and feel free to comment if something seems wrong to you.

Now I am working on [document ](https://docs.google.com/document/d/1fYSoD8RuQdWPjgnArK7pFtxbMqjrXZwbpoHTiYUTa4w/edit?usp=sharing) where I investigate how duplications between casks and formulae can be resolved. I thought that doc would be a nice idea not to flood in related issue and plan on sharing it there later for discussion as it is sometimes not perfectly clear for me how to tell cask-related app from homebrew-related.

Thank you,
Anastasia

**Comment by vitorgalvao** [comment]
Hello all, apologies for the delay, and thank you for the interest, @AnastasiaSulyagina! It took me so long to write here partly because I’m currently abroad (will get back near the end of your first interval, in the first document), and partly because I had too much to write.

I’ve read the documents. I was actually going to mention de-duplication is only tangential to this and wouldn’t need to be included in this merger, but then I saw your second document and really liked what you’ve done so far and the raised points.

---

#### De-duplication

First, lets talk de-duplication. There are some overlapping considerations, and we need to define them in order of importance to have consistent enforceable rules. I believe this to be even more important than purely following logic. What I mean by this is if we do it in a “it makes sense to me this like this” way, the rule will be applied inconsistently *and what seems logical to a maintainer might be illogical to a user*. If, however, we have clear rules we can point users to, that gives us dependable consistency, which I believe to be of the utmost importance. We’ve been working in this way in HBC, with positive results and only the occasional user confusion (that we can easily clarify via the documentation). Exceptions will always need to occur, but they should be kept in check.

Current situation: HB takes OSS-only CLI-only apps, with some exceptions for GUI OSS apps. HBC takes OSS and proprietary apps, both CLI and GUI. As a small example of why the division isn‘t clear-cut, we can’t simply say all GUI apps should be moved to HBC, because some can’t be in HBC — [`mpv`](https://github.com/mpv-player/mpv/releases), for example, hasn’t been providing GUI app bundles, which means HBC is behind HB for that app.

We need then to figure out how to best divvy-up casks and formulas. This is mostly non-technical work, and **it isn’t strictly necessary to achieve the objective of integrating both projects**, since we’ll still have separate repos for casks and formulas (so de-duplication can progress even after we merge). However, you’re doing stellar work in identifying the differences between the current duplicates, and if we finish the list, we’ll have a better way to decide how things should be divided.

Honestly, the more I think about this, the more I think we should scrap our custom DSL [and integrate apps by extending Homebrew’s DSL](https://github.com/caskroom/homebrew-cask/issues/14384#issuecomment-147761802) (warning: humungous post) but I’m digressing.

---

On to the rest of your proposal/questions.

#### Changing HBC’s behaviour

It would be nice if the [behaviour change](https://github.com/caskroom/homebrew-cask/issues/13201) could be in place for this, yes, but I suppose it *might* not be strictly necessary. However, it would be insanely useful if it were already in place, since otherwise a lot of the logic of merging both projects might need to be redone, and that would be a waste of resources. @jawshooah does most of the reviews to the core changes (since he’s the best at ruby of the team), but there are occasionally periods when he can’t spend as much time to the project, due to external circumstances. Conversely, other times he’s a beast of contributions, reviewing and enhancing left and right. We have [a PR in progress for the behaviour change](https://github.com/caskroom/homebrew-cask/pull/13966), so if you’d like to take a look at reviewing/fixing what’s missing, that’s the place to go.

#### Separating casks from the core

Probably what really needs to be done in advance is [separating casks from the core](https://github.com/caskroom/homebrew-cask/issues/1155), especially since Homebrew finally also finally did it itself so recently. Congratulations, by the way! As a user, it was a beautifully seamless transition.

---

Apologies again for the time it took me to reply (and the post’s length). Feel free to hit me up with anything I forgot to mention or was unclear on. And again, thank you for tackling this.

**Comment by AnastasiaSulyagina** [comment]
Thank you for such a profound answer, @vitorgalvao !
I will take a look at what is already done about behavior change and in May I am going to start working towards separating casks from the core. Also I will put a link to deduplication doc to the related issue so that it could be discussed and proceeded with in parallel with the necessary changes.

**Comment by mikemcquaid** [comment]
@AnastasiaSulyagina was accepted to GSoC 2016 for this project 🎉
